File: apps/mobile/lib/core/services/email_sync_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

part 'email_sync_service.g.dart';

class EmailSyncService {
  final FirebaseAuth _auth;
  final FirebaseFirestore _firestore;

  EmailSyncService(this._auth, this._firestore);

  /// Checks if the user's email in Firestore matches their Firebase Auth email
  /// and updates it if necessary. This should be called after successful login.
  Future<void> syncUserEmailIfNeeded() async {
    try {
      final currentUser = _auth.currentUser;
      if (currentUser == null) {
        // User not logged in, skip sync
        return;
      }

      final uid = currentUser.uid;
      final firebaseEmail = currentUser.email;

      // Skip if Firebase user doesn't have an email
      if (firebaseEmail == null || firebaseEmail.isEmpty) {
        return;
      }

      // Check if user document exists
      final docRef = _firestore.collection('users').doc(uid);
      final docSnapshot = await docRef.get();

      if (!docSnapshot.exists) {
        // Document doesn't exist, skip sync
        return;
      }

      final docData = docSnapshot.data();
      if (docData == null) {
        // Document data is null, skip sync
        return;
      }

      final firestoreEmail = docData['email'] as String?;

      // Check if emails are different
      if (firestoreEmail != firebaseEmail) {
        // Update the email in Firestore
        await docRef.update({
          'email': firebaseEmail,
          'lastEmailSync': FieldValue.serverTimestamp(),
        });
      }
    } catch (e) {
      // Silently handle errors to avoid breaking the login flow
      // In production, you might want to log this error
      print('Email sync failed: $e');
    }
  }

  /// Checks if the user is an Apple user without an email address
  /// This helps identify users affected by the missing email scope issue
  Future<bool> isAppleUserWithoutEmail() async {
    try {
      final currentUser = _auth.currentUser;
      if (currentUser == null) return false;

      // Check if user has Apple provider
      final hasAppleProvider = currentUser.providerData
          .any((provider) => provider.providerId == 'apple.com');

      if (!hasAppleProvider) return false;

      // Check if user has no email or empty email
      final hasNoEmail =
          currentUser.email == null || currentUser.email!.isEmpty;

      return hasNoEmail;
    } catch (e) {
      return false;
    }
  }

  /// Gets the user's authentication providers for debugging/analytics
  List<String> getUserProviders() {
    final currentUser = _auth.currentUser;
    if (currentUser == null) return [];

    return currentUser.providerData
        .map((provider) => provider.providerId)
        .toList();
  }

  /// Checks if user needs email collection (legacy Apple users)
  Future<bool> shouldPromptForEmailCollection() async {
    try {
      final currentUser = _auth.currentUser;
      if (currentUser == null) return false;

      // Only prompt Apple users without emails
      if (!await isAppleUserWithoutEmail()) return false;

      // Check if we've already prompted this user (to avoid spam)
      final docRef = _firestore.collection('users').doc(currentUser.uid);
      final docSnapshot = await docRef.get();

      if (!docSnapshot.exists) return false;

      final docData = docSnapshot.data();
      final hasBeenPrompted = docData?['emailCollectionPrompted'] == true;

      // Don't prompt if we've already prompted them
      return !hasBeenPrompted;
    } catch (e) {
      return false;
    }
  }

  /// Marks that the user has been prompted for email collection
  Future<void> markEmailCollectionPrompted() async {
    try {
      final currentUser = _auth.currentUser;
      if (currentUser == null) return;

      final docRef = _firestore.collection('users').doc(currentUser.uid);
      await docRef.update({
        'emailCollectionPrompted': true,
        'emailCollectionPromptedAt': FieldValue.serverTimestamp(),
      });
    } catch (e) {
      // Silently handle errors
    }
  }
}

@Riverpod(keepAlive: true)
EmailSyncService emailSyncService(Ref ref) {
  return EmailSyncService(
    FirebaseAuth.instance,
    FirebaseFirestore.instance,
  );
}

File: apps/ios/ios/Core/Services/EmailSyncService.swift
import Foundation
import FirebaseAuth
import FirebaseFirestore

/// Syncs Firebase Auth email to Firestore and handles Apple users without email
/// Ported from: apps/mobile/lib/core/services/email_sync_service.dart
@Observable
@MainActor
final class EmailSyncService {

    private let db = Firestore.firestore()

    /// Syncs the user's Firebase Auth email to their Firestore document if different
    func syncUserEmailIfNeeded() async {
        guard let currentUser = Auth.auth().currentUser,
              let firebaseEmail = currentUser.email, !firebaseEmail.isEmpty else {
            return
        }

        let uid = currentUser.uid
        let docRef = db.collection("users").document(uid)

        do {
            let snapshot = try await docRef.getDocument()
            guard snapshot.exists, let data = snapshot.data() else { return }

            let firestoreEmail = data["email"] as? String

            if firestoreEmail != firebaseEmail {
                try await docRef.updateData([
                    "email": firebaseEmail,
                    "lastEmailSync": FieldValue.serverTimestamp(),
                ])
            }
        } catch {
            print("[EmailSyncService] Email sync failed: \(error.localizedDescription)")
        }
    }

    /// Checks if the user is an Apple user without an email address
    func isAppleUserWithoutEmail() -> Bool {
        guard let user = Auth.auth().currentUser else { return false }

        let hasAppleProvider = user.providerData.contains { $0.providerID == "apple.com" }
        guard hasAppleProvider else { return false }

        return user.email == nil || user.email?.isEmpty == true
    }

    /// Gets the user's auth provider IDs
    func getUserProviders() -> [String] {
        guard let user = Auth.auth().currentUser else { return [] }
        return user.providerData.map(\.providerID)
    }

    /// Checks if user should be prompted for email collection (legacy Apple users)
    func shouldPromptForEmailCollection() async -> Bool {
        guard let user = Auth.auth().currentUser else { return false }
        guard isAppleUserWithoutEmail() else { return false }

        do {
            let doc = try await db.collection("users").document(user.uid).getDocument()
            guard doc.exists, let data = doc.data() else { return false }
            return data["emailCollectionPrompted"] as? Bool != true
        } catch {
            return false
        }
    }

    /// Marks that the user has been prompted for email collection
    func markEmailCollectionPrompted() async {
        guard let user = Auth.auth().currentUser else { return }
        try? await db.collection("users").document(user.uid).updateData([
            "emailCollectionPrompted": true,
            "emailCollectionPromptedAt": FieldValue.serverTimestamp(),
        ])
    }
}
